### ISSUE TITLE
[css-typed-om-1] Should css typed om support the same simplifications as css-values-4?

### ISSUE THREAD

**Issue by danielsakhapov** [w3c/csswg-drafts#9451]
So, the question is should https://drafts.css-houdini.org/css-typed-om-1/ support the same simplifications (like distributing multiplication over sum and sorting the dimensions alphabetically) as css-values-4? Like https://drafts.csswg.org/css-values-4/#calc-simplification and https://drafts.csswg.org/css-values-4/#sort-a-calculations-children?

**Comment by andruud** [comment]
Reading the specs, this is what it looks like to me:

 - cssom/css-values _simplifies_ during _parsing_.  (E.g. distribution, but not sorting).
 - cssom/css-values _sorts by unit_ during _serialization_. 

However:

- css-typed-om _does not_ simplify during "create an internal representation".
- css-typed-om _does not_ sort by unit during reification.

So this means:

 - Any internal representation you create from a `CSSMathValue` is the same thing you get when it's reified back. It's not simplified nor sorted. (Specified values).
 - An internal representation produced from _parsing_, will be simplified (but not sorted), and so it will reify as the simplified structure (again, unsorted).
 - An internal representation produced from a `CSSMathValue` will be _sorted_ when serialized (cssom/css-values), but not _simplified_.

This is a mess, but I think we can make it more sane if we do one of the following:

 - Move all sorting and simplification to _parsing_, or
 - Move all sorting and simplification to _serialization_. (Maybe better? https://github.com/w3c/csswg-drafts/issues/9437)

In any case, preserving `CSSMathValue`-produced values as-is seems like nice behavior worth maintaining.

cc @tabatkins

**Comment by tabatkins** [comment]
The behavior is well-defined across the specs. Simplification occurs at parse-time, and when transitioning between some value stages (specified->computed, computed->used).

Typed OM represents a value at one of these stages; whatever the form is at that moment is what it should be reflected as in the Typed OM. If you build one yourself, no simplification happens on the object graph (but once you assign it to a property, the internal value will be simplified as normal).

(The "create an internal representation" algo in Typed OM is a bit too handwavey, and could use some editorial tweaking, tho. I'll untag this from Values and leave it as a Typed OM issue.)

**Comment by andruud** [comment]
> but once you assign it to a property, the internal value will be simplified as normal

OK, it sounds like you're really expecting the creation of an internal value (whether that comes from parsing or typed-om or something else) to invoke [simplify a calculation tree](https://drafts.csswg.org/css-values-4/#simplify-a-calculation-tree). Currently, only parsing invokes that, so we're bypassing it with typed-om (for specified values).

We can currently also bypass [sorting](https://drafts.csswg.org/css-values-4/#sort-a-calculations-children) if we reify via typed-om and serialize the CSSStyleValue. Is that intentional?

**Comment by Mstanic05** [comment]
Can you assign this to me!

**Comment by emilio** [comment]
I agree that the current state of things is a mess. We found this while implementing TypedOM in gecko and would rather sort and simplify always at the same time. @andruud any objections from chromium? There are some tests that currently depend on this behavior.

**Comment by andruud** [comment]
> gecko [...] would rather sort and simplify always at the same time. @andruud any objections from chromium?

No objections to that.

**Comment by tabatkins** [comment]
Yeah, simplification/sorting isn't defined to be a parse-only operation; I think it would be best to always sort/simplify together like Gecko.
